tools/degreeworks-scraper/src/components/Scraper.ts

```diff
@@ -0,0 +1,190 @@
+import { jwtDecode } from "jwt-decode";
+import type { JwtPayload } from "jwt-decode";
+
+import type { Program } from "../types";
+
+import { AuditParser, DegreeworksClient } from ".";
+
+export class Scraper {
+  private ap!: AuditParser;
+  private dw!: DegreeworksClient;
+
+  private degrees: Map<string, string> | undefined = undefined;
+  private majorPrograms: Set<string> | undefined = undefined;
+  private minorPrograms: Set<string> | undefined = undefined;
+
+  private done = false;
+  private parsedMinorPrograms: Map<string, Program> | undefined = undefined;
+  private parsedUgradPrograms: Map<string, Program> | undefined = undefined;
+  private parsedGradPrograms: Map<string, Program> | undefined = undefined;
+  private parsedSpecializations: Map<string, Program> | undefined = undefined;
+  private degreesAwarded: Map<string, string> | undefined = undefined;
+
+  private constructor() {}
+
+  private async scrapePrograms(school: string, degrees: Set<string>) {
+    if (!this.majorPrograms) throw new Error("majorPrograms has not yet been initialized.");
+    const ret = new Map<string, Program>();
+    for (const degree of degrees) {
+      for (const majorCode of this.majorPrograms) {
+        const audit = await this.dw.getMajorAudit(degree, school, majorCode);
+        if (!audit) {
+          console.log(
+            `Requirements block not found (majorCode = ${majorCode}, degree = ${degree})`,
+          );
+          continue;
+        }
+        if (ret.has(audit.title)) {
+          console.log(
+            `Requirements block already exists for "${audit.title}" (majorCode = ${majorCode}, degree = ${degree})`,
+          );
+          continue;
+        }
+        ret.set(audit.title, this.ap.parseBlock(`${school}-MAJOR-${majorCode}-${degree}`, audit));
+        console.log(
+          `Requirements block found and parsed for "${audit.title}" (majorCode = ${majorCode}, degree = ${degree})`,
+        );
```

Comment:
Sleep after the audit has been fetched and process in its entirety.

```suggestion
      for (const majorCode of this.majorPrograms) {
        const audit = await this.dw.getMajorAudit(degree, school, majorCode);
        if (!audit) {
          console.log(
            `Requirements block not found (majorCode = ${majorCode}, degree = ${degree})`,
          );
          continue;
        }
        if (ret.has(audit.title)) {
          console.log(
            `Requirements block already exists for "${audit.title}" (majorCode = ${majorCode}, degree = ${degree})`,
          );
          continue;
        }
        ret.set(audit.title, this.ap.parseBlock(`${school}-MAJOR-${majorCode}-${degree}`, audit));
        console.log(
          `Requirements block found and parsed for "${audit.title}" (majorCode = ${majorCode}, degree = ${degree})`,
        );
        
        // For example
        console.log('Done processing audit, now sleeping before processing next...')
        this.sleep(1000)
```